src/hotspot/share/gc/shared/referenceProcessor.cpp

```diff
@@ -264,6 +264,7 @@ void DiscoveredListIterator::remove() {
 }
 
 void DiscoveredListIterator::make_referent_alive() {
+  assert(referent() != nullptr, "precondition");
```

Comment:
This assert is not correct.  Preclean runs concurrently with the mutator.  The mutator could clear the reference between the time it is accessed by the list iterator and the call to make_referent_alive.  Looking at this also makes me wonder if the keep-alive function is safe against a racing clear by the mutator.

Reply:
Answering my own question. Only G1 currently uses precleaning, and it looks fine wrto a concurrent clear. Indeed, I think it's an expensive nop. I suspect (though haven't thought about it carefully yet) that any GC that uses ReferenceProcessor would do nothing there, and I'm not sure preclean actually needs the keep-alive and complete-gc closures.

Reply:
My thought was that it can prevent the scenario of unintentionally introducing an expensive nop (marking null referent). Removed now.